### PR TITLE
Decouples light-sync state from chain spec

### PR DESCRIPTION
**Pull request by bkchr**
This decouples the light-sync state from chain spec. First, the
light-sync state currently only works with BABE+Grandpa, so not
all *Substrate* based chains can use this feature. The next problem was
also that this pulled the `sc-consensus-babe` and `sc-finality-grandpa`
crate into `sc-chain-spec`.

If a chain now wants to support the light-sync state, it needs to add
the `LightSyncStateExtension` to the chain spec as an extension. This is
documented in the crate level docs of `sc-sync-state-rpc`. If this
extension is not available, `SyncStateRpc` fails at initialization.

polkadot companion: https://github.com/paritytech/polkadot/pull/3568

Fixes: https://github.com/paritytech/substrate/issues/8886

### CODE REVIEW

File: client/consensus/babe/src/lib.rs [hunk 1]
```diff
@@ -129,7 +129,7 @@ pub use sp_consensus_babe::{
 	},
 	AuthorityId, AuthorityPair, AuthoritySignature, BabeApi, BabeAuthorityWeight,
 	BabeEpochConfiguration, BabeGenesisConfiguration, ConsensusLog, BABE_ENGINE_ID,
-	VRF_OUTPUT_LENGTH,
+	VRF_OUTPUT_LENGTH, BabeBlockWeight,
```
Comment (andresilva) [inline client/consensus/babe/src/lib.rs:132]: You need to configure your editor to cargo fmt now 😏.

**Review by andresilva — APPROVED** [review]

**Comment by arkpar** [comment]
What is the "light sync state" used for? Looks like the code that was supposed to use it was never written? Why not just remove it?

Also, the light client does not really need to have epochs and authorities in the chain spec. All that's needed is a trusted finalized block hash. The rest should be recoverable from the state.

**Comment by bkchr** [comment]
> What is the "light sync state" used for? Looks like the code that was supposed to use it was never written? Why not just remove it?
> 
> Also, the light client does not really need to have epochs and authorities in the chain spec. All that's needed is a trusted finalized block hash. The rest should be recoverable from the state.

It used by the light clients as a starting state. @tomaka should be better at explaining for what this is actually required.

**Comment by andresilva** [comment]
It's used by smoldot to start syncing from this checkpoint, on substrate only the code to export the sync state exists. This was written before warp sync was implemented on smoldot so maybe it's not super relevant anymore.

> Also, the light client does not really need to have epochs and authorities in the chain spec. All that's needed is a trusted finalized block hash. The rest should be recoverable from the state.

I think when this was written there was no easy way to get this data from the runtime, I agree that now if we still want to have a checkpoint for light clients to start syncing we only need to provide finalized block hash, we can even start warp syncing from such a block. Since this is used by smoldot currently I'll defer to @tomaka on what to do right now.

**Comment by tomaka** [comment]
The light sync state was "designed" and added before Grandpa warp sync existed.

At the time, we were also trying as much as possible to avoid having to do runtime calls in the light client (as this makes the light client binary size much larger than otherwise, and requires downloading the runtime code from the network). For this reason, retrieving the Babe epoch and finalized authorities from the state wasn't really a good solution.

Nowadays, yes we probably just need a finalized block hash. However, knowing the list of Grandpa authorities of this finalized block is necessary in order to verify the warp syncing. This list could also be retrieved from the network, but this adds some latency and most importantly we run the risk that all the nodes of the network have already pruned the block and are incapable of giving us that list.

**Comment by bkchr** [comment]
So, you would like to keep it @tomaka?

**Comment by tomaka** [comment]
Right now, this light sync state contains what a light client needs in order to start syncing from a certain block.

We could remove the Babe-related stuff from the light sync state, but keep the Grandpa-related stuff, in which case the light sync state would contain what a light client needs in order to warp sync from a certain block.

If we remove also the Grandpa-related stuff and keep only a finalized block header or hash, then we risk that everyone on the network has pruned the block, and/or that we make lots of useless network requests trying to find a node that hasn't. This is too much removal IMO.

Smoldot can accommodate any of the three, I don't care, but I'd be leaning towards removing the Babe stuff.

**Comment by bkchr** [comment]
Okay, then I'm going with keeping it as is for now.

**Review by arkpar — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Comment by apopiak** [comment]
Could it be that a Cumulus companion is missing for this?

**Pull request merged**